### PR TITLE
Update refresh strategy to use a separate refresh token

### PR DESCRIPTION
**Pull request by evoactivity**
As discussed in this issue #162 I've made the library use a separate token to refresh the JWT.

I've updated the tests and they all pass but someone should really check what I've done, testing is not my strong point.

I've ran this on my app and confirmed it refreshes the token as expected.

### CODE REVIEW

**Comment by eluciano11** [comment]
@evoactivity thank you for your work! 

I've looked at the test and the implementation and it looks good. I think that we should wait until @jpadilla can take a look at this before we merge the PR. This is a breaking change so we'll need to update the docs and demo too. We can do that after @jpadilla takes a look.

**Comment by evoactivity** [comment]
@eluciano11 No problem :)

I was wondering about the docs/demos, but decided against updating them until this had been reviewed incase there are any changes needed.

File: addon/authenticators/jwt.js
Comment (alechirsch) [inline addon/authenticators/jwt.js]: Don't we want to check if the refresh token has expired rather than if the authentication token has expired? That is my understanding of refresh tokens, they out last the authentication tokens.

File: addon/authenticators/jwt.js
Comment (alechirsch) [inline addon/authenticators/jwt.js:122]:
```javascript
if (Ember.isEmpty(expiresAt)) {
         // Fetch the expire time from the token data since `expiresAt`
         // wasn't included in the data object that was passed in.
         const tokenData = this.getTokenData(token);
 
         expiresAt = tokenData[this.tokenExpireName];
         if (Ember.isEmpty(expiresAt)) {
           return resolve(data);
         }
       }
 
       if (expiresAt > now) {
```

Shouldn't we be checking if the refresh token `expiresAt > now` rather than the auth token `expiresAt > now`?

With the way it currently is, if the auth token expired but the refresh token has not expired, it would not attempt to refresh.
I think we would want to attempt to refresh if the refresh token is still valid.

Comment (evoactivity) [reply addon/authenticators/jwt.js:122]: My API's refresh token library creates the token like so `$this->refreshToken = bin2hex(openssl_random_pseudo_bytes(64));`, so the refresh token is not just another JWT with an expires property to be decoded. Looking at the auth0 page on refresh tokens it looks like they're implementing them in a similar way https://auth0.com/learn/refresh-tokens/

If the auth token is expired that's exactly when this tries to refresh using the refresh token, so I'm not entirely sure what you mean by that?

Comment (alechirsch) [reply addon/authenticators/jwt.js:122]: From what I am seeing, if the auth token is expired when it enters the restore function then it will reject on line 132
```
 reject(new Error('token is expired'));
```

Comment (evoactivity) [inline addon/authenticators/jwt.js]: Ahh, of course I see what you mean now. Yeah, you're right, if your token is already expired it won't be refreshed.

I think the change here would be if it's expired to attempt a refresh, if the server rejects the refresh token then unauth the session so they are sent to the login form.

Comment (evoactivity) [reply addon/authenticators/jwt.js:122]: Oh yes, I see what you mean now, my mistake.

I guess if the token is expired and we have `refreshAccessTokens` set to true it should also attempt to refresh the token there as well, and if the server rejects the refresh token then the user is unauthorised sending them back to the login form.

Comment (alechirsch) [reply addon/authenticators/jwt.js:122]: Yes, sorry for the confusion.

**Review by jpadilla — APPROVED** [review]

**Comment by jpadilla** [comment]
@evoactivity At a first pass, this lgtm. Thanks for putting this together.
@eluciano11 is there anything we'd need after this, docs and dummy app?

I'm fine introducing this as a new major release, but would also love to sneak in updating everything else including to the latest ESA.

File: addon/authenticators/jwt.js
Comment (alechirsch) [reply addon/authenticators/jwt.js:122]: @evoactivity Do you have time to implement the fix for this? I am considering forking off, I will be trying to use this feature in an app for this week.

Comment (evoactivity) [reply addon/authenticators/jwt.js:122]: @alechirsch I will be looking at this again maybe later today or tomorrow.

Comment (evoactivity) [reply addon/authenticators/jwt.js:122]: @alechirsch I've updated this to attempt a refresh if the JWT is expired. If the server rejects the request the session is invalidated.

Comment (alechirsch) [reply addon/authenticators/jwt.js:122]: @evoactivity Awesome, thanks for your work on this, it is perfect timing.

**Comment by evoactivity** [comment]
@jpadilla no worries :) I've just pushed a small update to address the expiry issue pointed out by @alechirsch 

The docs will need updating to explain the new `refreshTokenPropertyName` option, if people want to use the current strategy they should be able to simply give that the same name as `tokenPropertyName`.

The demo app should probably be updated to show this being used, the demo server will need to return two tokens now, the JWT and a standard refresh token.

**Review by alechirsch — APPROVED** [review]

**Comment by eluciano11** [comment]
@jpadilla @evoactivity yup, updated the docs and the demo app but it's not using the two tokens it's falling back to the old strategy.

**Comment by alechirsch** [comment]
I just did some testing in my own application, there seems to be an issue with old timeouts not being removed. I will do more testing to find out what is really going on.

**Comment by alechirsch** [comment]
@evoactivity From what I am finding if you log in then refresh the page to make it call restore, it will refresh the token, and then call restore again exponentially. Meaning it calls restore once, then twice, then 4 times, then 8. Something about `resolve(this.refreshAccessToken(refreshToken));` is making it call restore twice for every time it is called.

Restore should be called only once, when the page is refresh or opened in a new tab. I am unsure why this is happening at the moment.

**Comment by evoactivity** [comment]
@alechirsch I've not seen that behaviour at all in my own app.

What sort of expiry time do you have on your JWT? Also, what leeway do you have set?

**Comment by alechirsch** [comment]
For testing purposes I set it to 60 seconds, with 55 second leeway. Maybe that is the problem?

**Comment by evoactivity** [comment]
I've been testing with my JWT set to 1 minute and my refresh token set to 2 minutes and my leeway set to 0.

I'm not sure why a 5 second expiry would call exponential calls to refreshAccessToken though.

**Comment by alechirsch** [comment]
What's odd is that the restore function is being called more than once.

**Comment by alechirsch** [comment]
I blew away my node_modules and reinstalled, set my expire to 10 minutes and leeway to 60 seconds. I am running into the same problem, the restore function is being called in an infinite loop somehow. It works great if the restore function rejects the first time it is hit, however if it resolves, it calls itself again. It hogs up the CPU at 100% as well.

**Comment by alechirsch** [comment]
I have traced it back to this file `ember-simple-auth/instance-initializers/setup-session-restoration`.

The beforeModel hook is being called on an infinite loop, which is in turn calling restore. This might be a problem with my app, looking into it more.

**Comment by alechirsch** [comment]
@evoactivity It was definitely a problem with my application, the feature works as expected now.

**Comment by jpadilla** [comment]
This is the current server implementation that the dummy app uses https://github.com/jpadilla/ember-simple-auth-token-server

**Comment by alechirsch** [comment]
What is currently holding up this PR? I have been using this for the past few weeks in my apps and it works perfectly by the way.

**Comment by jpadilla** [comment]
@alechirsch thanks for bumping this again, great to know you've been using it for a while now. Since this is a breaking change, it'll be released on v3, but before doing so I would love to make sure we update ember-cli, ESA, etc. Thanks to all!

PS If anyone else wants to help with that before I do, please feel free to do so. Thanks!

**Comment by jpadilla** [comment]
@homu r+

**Comment by jpadilla** [comment]
@alechirsch @evoactivity I'm looking for help with maintenance, would any of you be up to it?

**Comment by alechirsch** [comment]
@jpadilla I can help out with pull requests and issues

**Comment by jpadilla** [comment]
@alechirsch that's so great, just sent you an invite, thank you so much!

**Comment by evoactivity** [comment]
@jpadilla I'm too busy to offer much help at the moment but I'd be interested in helping when things settle down a bit for me. That's probably a month or two away though.

**Comment by jpadilla** [comment]
@evoactivity understandable! Feel free to ping back whenever you have a few cycles free, thanks!

**Comment by knownasilya** [comment]
This is a breaking change, because if the refreshToken name isn't set it defaults to `refresh_token`, which used to be the token name. If the property isn't set, it should default to the token name.

**Comment by jpadilla** [comment]
Thank you all again! Just released [v3.0.0](https://github.com/jpadilla/ember-simple-auth-token/releases/tag/v3.0.0) 🎉

**Pull request merged**
